### PR TITLE
Fix interlace check for H.264 MBAFF coded MP4 files

### PR DESCRIPTION
**Pull request by orryverducci**
**Changes**
FFprobe does not return a field order for some interlaced H.264 files using MBAFF coding in an MP4 container. As a result Jellyfin does not detect that these files are interlaced.

To work around this, it can be determined that the video is interlaced by comparing the codec time base to the video frame rate. If the time base duration is half of a frame's duration, the video is interlaced.

This change adds this check for MP4 H.264 files that don't return a field order, fixing interlace detection for them.

### CODE REVIEW

File: MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs [hunk 1]
```diff
@@ -746,6 +741,19 @@ private MediaStream GetMediaStream(bool isAudio, MediaStreamInfo streamInfo, Med
...
                 stream.RealFrameRate = GetFrameRate(streamInfo.RFrameRate);
 
+                // Some interlaced H.264 files in mp4 containers using MBAFF coding aren't flagged as being interlaced by FFprobe,
+                // so for H.264 files we also check if the codec timebase duration is half the reported frame rate duration to
+                // determine if the file is interlaced
+                bool videoInterlaced = !string.IsNullOrWhiteSpace(streamInfo.FieldOrder)
+                    && !string.Equals(streamInfo.FieldOrder, "progressive", StringComparison.OrdinalIgnoreCase);
+                bool h264MbaffCoded = string.Equals(stream.Codec, "h264", StringComparison.OrdinalIgnoreCase)
+                    && string.IsNullOrWhiteSpace(streamInfo.FieldOrder)
+                    && 1f / (stream.AverageFrameRate * 2) == GetFrameRate(stream.CodecTimeBase);
```
Comment (GodTamIt) [inline MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:751]: Using floating point equality here will [probably not be accurate](https://bitbashing.io/comparing-floats.html) all of the time.

Is there another way to do this calculation?

Comment (orryverducci) [reply MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs]: Unfortunately there isn't really any other way to do the calculation, as FFprobe returns fractional values.

However I have changed it to round the values, which should hopefully mitigate this issue.

**Review by nyanmisaka — APPROVED** [review]
LGTM

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs [hunk 2]
```diff
@@ -724,6 +719,23 @@ private MediaStream GetMediaStream(bool isAudio, MediaStreamInfo streamInfo, Med
...
+                // so for H.264 files we also calculate the frame rate from the codec time base and check if it is double the reported
+                // frame rate (both rounded to the nearest integer) to determine if the file is interlaced
+                float roundedTimeBaseFPS = MathF.Round(1 / GetFrameRate(stream.CodecTimeBase) ?? 0);
+                float roundedDoubleFrameRate = MathF.Round(stream.AverageFrameRate * 2 ?? 0);
+
+                bool videoInterlaced = !string.IsNullOrWhiteSpace(streamInfo.FieldOrder)
+                    && !string.Equals(streamInfo.FieldOrder, "progressive", StringComparison.OrdinalIgnoreCase);
+                bool h264MbaffCoded = string.Equals(stream.Codec, "h264", StringComparison.OrdinalIgnoreCase)
+                    && string.IsNullOrWhiteSpace(streamInfo.FieldOrder)
+                    && roundedTimeBaseFPS == roundedDoubleFrameRate;
```
Comment (Bond-009) [inline MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:732]: Don't like this

**Comment by dmitrylyzo** [comment]
This video is detected as interlaced, but it is not.

https://user-images.githubusercontent.com/56478732/152567222-ae38aa1e-bc4e-4c68-b8ed-93f16aaaf5fd.mp4

```cs
videoInterlaced: false
h264MbaffCoded: true
roundedTimeBaseFPS: 50 // '1/50'
roundedDoubleFrameRate: 50 // 25 * 2
```

**Comment by gaudat** [comment]
Hello @dmitrylyzo I stumbled upon this issue when finding why MP4 test videos generated by ffmpeg is recognized as interlaced in Jellyfin.

I used the following command to generate the video:
```
/c/Users/anon/Apps/ffmpeg/bin/ffmpeg -re -f lavfi -i "smptehdbars=rate=30:size=1920x1080" \
-f lavfi -i "sine=frequency=400:sample_rate=48000" \
-vf setparams="field_mode=prog",drawtext="text='1080 30fps h264 mp4 %{pts\:hms}':rate=30:x=(w-tw)/2:y=(h-lh)/2:fontsize=48:fontcolor=white:box=1:boxcolor=black" \
 -to 10 -c:v h264 -profile:v baseline -pix_fmt yuv420p -preset ultrafast -tune zerolatency -crf 28 -c:a aac testvid.mp4
```

This video along with the 2 of yours do not have a "fiel" atom in the file structure. In this case ffprobe will report the video as neither interlaced nor progressive.

It is unlucky that the timebase duration is the same as half of the video frame duration in our videos. I have yet to find a way to change it in ffmpeg.

One has to bend ffmpeg in weird ways to make it spit out a interlaced H264 video as shown in this SO question... [Link](https://superuser.com/questions/1344442/ffmpeg-how-to-create-an-interlaced-h-264-video)
ffmpeg's "interlace detection" is unreliable. It detects @dmitrylyzo 's interlaced video as mostly progressive, while my supposedly interlaced video is undetermined.

It is pretty rare to see interlaced video in this age of digital streaming. I think the last device I used that gives out interlaced video is a camcorder made in 2011. Is it time to change the default from interlaced to progressive, or to whitelist modern codecs (H264/H265/VPx) such that these are progressive by default?

**Pull request merged**